server/routes/charts.js

```diff
@@ -42,55 +43,72 @@ router.route('/charts/new').get(ensureAuthenticated, async (req, res) => {
   }
 })
 
-router.route('/charts/:chart_id').get(ensureAuthenticated, async (req, res) => {
-  try {
-    const { dataDb, appDb } = req.app.locals
-    const { chart_id } = req.params
-    const { userAccess } = req.session
-    const parsedQueryParams = qs.parse(req.query)
-    const {
-      chart: { title, description, fieldLabelValueMap, owner },
-      dataBySite,
-      labels,
-      studyTotals,
-      filters,
-    } = await graphDataController(
-      dataDb,
-      userAccess,
-      chart_id,
-      parsedQueryParams
-    )
-    const user = userFromRequest(req)
-    const chartOwner = await appDb.collection(collections.users).findOne(
-      { uid: owner },
-      {
-        projection: {
-          _id: 0,
-          display_name: 1,
-          icon: 1,
-          uid: 1,
-        },
+router
+  .route('/charts/:chart_id')
+  .get(ensureAuthenticated, async (req, res, next) => {
+    try {
+      const { dataDb, appDb } = req.app.locals
+      const { chart_id } = req.params
+      const { userAccess } = req.session
+      const parsedQueryParams = qs.parse(req.query)
+      const {
+        chart: { title, description, fieldLabelValueMap, owner },
+        dataBySite,
+        labels,
+        studyTotals,
+        filters,
+        graphTable,
+      } = await graphDataController(
+        dataDb,
+        userAccess,
+        chart_id,
+        parsedQueryParams
+      )
+      const user = userFromRequest(req)
+      const chartOwner = await appDb.collection(collections.users).findOne(
+        { uid: owner },
+        {
+          projection: {
+            _id: 0,
+            display_name: 1,
+            icon: 1,
+            uid: 1,
+          },
+        }
+      )
+      const graph = {
+        chart_id,
+        dataBySite,
+        labels,
+        title: title,
+        description: description,
+        legend: legend(fieldLabelValueMap),
+        studyTotals,
+        filters,
+        chartOwner,
+        graphTable,
       }
-    )
-    const graph = {
-      chart_id,
-      dataBySite,
-      labels,
-      title: title,
-      description: description,
-      legend: legend(fieldLabelValueMap),
-      studyTotals,
-      filters,
-      chartOwner,
-    }
+      if (req.headers['content-type'] === 'text/csv') {
+        res.header('Content-Type', 'text/csv')
+
+        const csvService = new CsvService(formatGraphTableDataToCSV(graphTable))
+        csvService.toReadableStream().pipe(res)
```

Comment:
does this work to pipe the stream to the `res` object? Do we not need to call `res.write`?

Reply:
yes it does :). The response object from express handles writing the data from the readable stream under the hood, but we still need to subscribe to the events and handle the errors and the end of the stream.